app/controllers/ListSchemesController.scala

```diff
@@ -21,96 +21,215 @@ import config.FrontendAppConfig
+  private def schemeDetails(
+    listOfSchemes: ListOfSchemes
+  ): List[SchemeDetail] = {
     listOfSchemes.schemeDetail.getOrElse(List.empty[SchemeDetail])
+  }
 
-  def renderView(schemeDetails: List[SchemeDetail], numberOfSchemes: Int, pageNumber: Int, numberOfPages: Int)
-                (implicit hc: HeaderCarrier, request: OptionalDataRequest[AnyContent]): Future[Result] = {
-
-    minimalPsaConnector.getPsaNameFromPsaID(request.psaId.id).flatMap(_.map {
-      name =>
-        userAnswersCacheConnector.save(request.externalId, PSANameId, name).map {
-          _ =>
-            Ok(view(
-              schemes = schemeDetails,
-              psaName = name,
-              numberOfSchemes = numberOfSchemes,
-              pagination = pagination,
-              pageNumber = pageNumber,
-              pageNumberLinks = paginationService.pageNumberLinks(pageNumber, numberOfSchemes, pagination, numberOfPages),
-              numberOfPages = numberOfPages
-            ))
-        }
-    }.getOrElse {
-      Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad()))
-    })
+  private def renderView(
+    schemeDetails: List[SchemeDetail],
+    numberOfSchemes: Int,
+    pageNumber: Int,
+    numberOfPages: Int,
+    noResultsMessageKey: Option[String],
+    form: Form[_]
+  )(implicit hc: HeaderCarrier,
+    request: OptionalDataRequest[AnyContent]): Future[Result] = {
+    val status = if (form.hasErrors) BadRequest else Ok
+    minimalPsaConnector
+      .getPsaNameFromPsaID(request.psaId.id)
+      .flatMap(_.map {
+        name =>
+          userAnswersCacheConnector
+            .save(request.externalId, PSANameId, name)
+            .map { _ =>
+              status(
+                view(
+                  form,
+                  schemes = schemeDetails,
+                  psaName = name,
+                  numberOfSchemes = numberOfSchemes,
+                  pagination = pagination,
+                  pageNumber = pageNumber,
+                  pageNumberLinks = paginationService.pageNumberLinks(
+                    pageNumber,
+                    numberOfSchemes,
+                    pagination,
+                    numberOfPages
+                  ),
+                  numberOfPages = numberOfPages,
+                  noResultsMessageKey
+                )
+              )
+            }
+      }.getOrElse {
+        Future.successful(
+          Redirect(controllers.routes.SessionExpiredController.onPageLoad())
+        )
+      })
   }
 
-  def onPageLoad: Action[AnyContent] = (authenticate andThen getData).async {
-    implicit request =>
-      listOfSchemes.flatMap {
-        listOfSchemes =>
-          val numberOfSchemes: Int = schemeDetails(listOfSchemes).length
+  private val srnRegex = "^S[0-9]{10}$".r
+  private val pstrRegex = "^[0-9]{8}[A-Za-z]{2}$".r
+
+  private val filterSchemesBySrnOrPstr
+    : (String, List[SchemeDetail]) => List[SchemeDetail] =
```

Comment:
can we move the logic of search from controller to its own service?

Reply:
Is it worth it? It will be a very small service!

Reply:
but Controller is not very readable. so if we separate it out it might look better.

Reply:
I will do it. I guess it will be worthwhile once we come to do the fuzzy matching

Reply:
Sure, will make controller more readable. Will sort it out after stand up

Reply:
I have made these changes Surya, please check and merge if you OK with changes.